lib/adf_derive.py

```diff
@@ -0,0 +1,274 @@
+    """
+    Derive variables acccording to steps given here.  Since derivations will depend on the
+    variable, each variable to derive will need its own set of steps below.
+
+    Caution: this method assumes that there will be one time series file per variable
+
+    If the file for the derived variable exists, the kwarg `overwrite` determines
+    whether to overwrite the file (true) or exit with a warning message.
+
+    """
+
+    # Loop through derived variables
+    print(f"\t - deriving time series for {var}")
+
+    # Grab all required time series files for derived variable
+    constit_files = []
+    for constit in constit_list:
+        # Check if the constituent file is present, if so add it to list
+        if glob.glob(os.path.join(ts_dir, f"*.{constit}.*.nc")):
+            constit_files.append(glob.glob(os.path.join(ts_dir, f"*.{constit}.*"))[0])
+    # End for
+
+    # Check if all the necessary constituent files were found
+    if len(constit_files) != len(constit_list):
+        ermsg = f"\t   ** Not all constituent files present; {var} cannot be calculated. **\n"
+        ermsg += f"\t     Please remove {var} from 'diag_var_list' or find the "
+        ermsg += "relevant CAM files.\n"
+        print(ermsg)
+        if constit_files:
+            # Add what's missing to debug log
+            dmsg = f"derived time series for {case_name}:"
+            dmsg += f"\n\tneeded constituents for derivation of "
+            dmsg += f"{var}:\n\t\t- {constit_list}\n\tfound constituent file(s) in "
+            dmsg += f"{Path(constit_files[0]).parent}:\n\t\t"
+            dmsg += f"- {[Path(f).parts[-1] for f in constit_files if Path(f).is_file()]}"
+            self.debug_log(dmsg)
+        else:
+            dmsg = f"derived time series for {case_name}:"
+            dmsg += f"\n\tneeded constituents for derivation of "
+            dmsg += f"{var}:\n\t\t- {constit_list}\n"
+            dmsg += f"\tNo constituent(s) found in history files"
+            self.debug_log(dmsg)
+        # End if
+    else:
+        # Open a new dataset with all the constituent files/variables
+        ds = self.data.load_dataset(constit_files)
+        if not ds:
+            dmsg = f"derived time series for {case_name}:"
+            dmsg += f"\n\tNo files to open."
+            self.debug_log(dmsg)
+            return
+
+        # Grab attributes from first constituent file to be used in derived variable
+        attrs = ds[constit_list[0]].attrs
+
+        # create new file name for derived variable
+        derived_file = constit_files[0].replace(constit_list[0], var)
+
+        # Check if clobber is true for file
+        if Path(derived_file).is_file():
+            if overwrite:
+                Path(derived_file).unlink()
+            else:
+                msg = f"[{__name__}] Warning: '{var}' file was found "
+                msg += "and overwrite is False. Will use existing file."
+                print(msg)
```

Comment:
Should we `return` here?  Otherwise it seems like we are doing calculations for no reason.

Reply:
Yeah, good catch!